src/Util/EntityFieldQuery.php

```diff
@@ -0,0 +1,137 @@
+<?php
+
+/**
+ * @file
+ * Contains \Drupal\restful\Util\EntityFieldQuery.
+ */
+
+namespace Drupal\restful\Util;
+
+use Drupal\restful\Exception\ServerConfigurationException;
+
+class EntityFieldQuery extends \EntityFieldQuery implements EntityFieldQueryRelationalConditionsInterface {
+
+  /**
+   * The relational filters.
+   *
+   * @var array[]
+   */
+  protected $relationships = array();
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getRelationships() {
+    return $this->relationships;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function addRelationship(array $relational_filter) {
+    $this->relationships[] = $relational_filter;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function finishQuery($select_query, $id_key = 'entity_id') {
+    foreach ($this->getRelationships() as $relationship) {
+      // A relational filter consists of a chain of relationships and a value
+      // for a condition at the end.
+      // Relationships start with the entity base table.
+      $entity_info = entity_get_info($this->entityConditions['entity_type']['value']);
+      $entity_table = $entity_table_alias = $entity_info['base table'];
```

Comment:
the `base table` might not be the entity table - it might be the field table -- it depends on the order the query was built. I Had the unfortunate opportunity to [deal](https://github.com/Gizra/og/blob/7.x-2.x/og.module#L1370) with how an alterable EFQ for D7 is on OG ;)